csrc/scheduler/normalization_inner_outer_tma_ws.cpp

```diff
@@ -174,10 +174,30 @@ void getHeuristics(
   // ping-pong computations.
   ParallelType ws_pt =
       iop.bdimx > 128 ? ParallelType::TIDx : ParallelType::TIDy;
+  WarpSpecialized ws(ws_pt);
+  // This is a heuristic para, multiple independent computation warp groups
+  // is not supported yet. Only need to enable register sharing when there
+  // are more than 256 threads, otherwise each thread can use 255 registers
+  // which is already the max allowed number.
+  int64_t independent_computation_groups = 1;
+  int64_t computation_threads = iop.bdimx * independent_computation_groups;
+  int64_t total_threads = ws_padded_threads + computation_threads;
+  if (total_threads > 256) {
+    int64_t reg_per_thread = getRegPerThreadGivenThreadsPerSM(total_threads);
+    // Assume each padded threads keep [tma_branch_registers] registers and all
+    // others are moved to computation threads. The granularity is 8.
+    // [tma_branch_registers] is a tunable parameter,
+    int64_t tma_branch_registers = 32;
+    int64_t compute_branch_registers = reg_per_thread +
+        (reg_per_thread - tma_branch_registers) * ws_padded_threads /
+            computation_threads;
+    compute_branch_registers =
+        scheduler_utils::roundDownToN(compute_branch_registers, 8);
```

Comment:
Naive question: do we need to cap `compute_branch_registers` to 256?

Reply:
Not really. We only use this feature when there are more than 256 threads, unless we reduce the `tma_branch_registers` to 0, the `compute_branch_registers` will always smaller than 256.